Banjo new tricks

CSharpM7 opened the pull request:
[Roadmap](https://docs.google.com/document/d/1GqlacWyOCtATlnybKdRrQhaaeFxQ_chYMVs9BdUhmms/edit?usp=sharing)
[Required romf file](https://www.dropbox.com/s/zwtnyvuemv1ava2/buddy_motion_body.7z?dl=0) (place in hdr-dev-assets/fighter/buddy/motion/body)

Contains the following new moves:

- Nair "Pack Wack"
- Upsmash "Old Uptilt / "Shock Spring Kick"
- Uptilt "Buddy Upper

Contains the following changes:

- Upair: First Active Frame 34->30
- Breegull Blaster: Weak egg threshold 6->10, Weakest egg threshold: 12->20. A small effect added when eggs become weak

CSharpM7 committed:
init

CSharpM7 committed:
New Tricks Part One

SkewedAskew commented:
Love the new animations, and the moves themselves feel great to use! I just have a couple initial thoughts:
- On utilt, Banjo's back arm is intangible during the startup and active frames. Was this intentional or is the other arm meant to have it instead?
- For up smash, please add a sys_flash somewhere near Banjo's head on startup, and consider adding 0x1156ac182a (circular air impact vfx) on his foot

CSharpM7 commented:
> Love the new animations, and the moves themselves feel great to use! I just have a couple initial thoughts:
> 
> *   On utilt, Banjo's back arm is intangible during the startup and active frames. Was this intentional or is the other arm meant to have it instead?
> *   For up smash, please add a sys\_flash somewhere near Banjo's head on startup, and consider adding 0x1156ac182a (circular air impact vfx) on his foot

For his uptilt, his back arm is intangible because he stretches his hurtbox relatively far for the side of the move which should be considered disjointed during the initial frames. Let me add in those effects as well.

CSharpM7 committed:
Up Smash Effects

Added smash attack indicator at beginning and circle impact at apex of kick

hdr-balance-backroom commented:
Looking at the design document and considering you have already previously implemented these in your own mod, we see no reason not to include all four phases into this PR. However, we have some thoughts on certain things:

- Beak Bomb should use the same resources as Wonderwing, though you're welcome to keep the visual shift to red feathers. The flight params will probably also need fine-tuning but we'll need to see that in-game before making further decisions
- For now, please revert nair's animation and hitbox data. We like the animation very much but are unsure if we want to sacrifice what the current nair does. Needs further discussion
- Please revert uair's FAF and hitbox data but increase the damage from 7% to 9%
- If you are able, we'd like to see dash attack follow in DK's footsteps and be able to slide off edges. Consult WuBoy if you have questions on how to implement it
- Regarding the Phase 3 changes specifically, the damage buffs and decrease to Breegull Blaster's startup are fine, but we do not feel that anything else from that page needs to be applied

Ravenash0 commented:
Just by looking at everything in the fact that I've been playing this pull request for the past like 2 weeks not only do I think this is a extremely balanced version of banjo but the neutral air actually add some variety rather than the stale neutral air that he has in vanilla, I really hope it makes it in

CSharpM7 committed:
Beakbomb, Bayonet, Damage Buffs, Fair Nerf

CSharpM7 committed:
Hitbox and reset fixes

CSharpM7 commented:
# Part 2

[Updated romfs](https://www.dropbox.com/s/gnn768kzsoahmbt/buddy_motion_body.7z?dl=0)
These contain the remaining changes to the New Tricks suite of Banjo changes.

### Aerial Side Special (BeakBomb)
```diff
- Startup: 18 -> 22
- FAF: 25 -> 35
- FAF (Wall): 17->47 (can be jump cancelled or air dodge cancelled from frame 15 if Banjo did not hit a Shield)
- 7% Armor on frames 16-22
- Hitbox size decreased
- Kazooie Hurtbox active
- Self damage on landing (only when in the air for less than 25 frames)
- Damage: 18/13 -> 16/10
- Recoil on shield
- 7.5 second cooldown (cooldown speeds up when on ground)
! BKB: 66->76
! KBG: 72->70
+ No longer uses Gold Feathers
+ Airspeed: 4.5 -> 4.6
+ Shield Damage: 0->4
+ Guard-cancel during dash
+ Trajectory control (on startup)
```

## Breegull Blaster (Neutral Special Stance)
When inputting a smash attack during this stance, you will now perform an (angled) side tilt.

## BackAir
Back air gets its vanilla KBG values back. While having a multihit combo move helps, the HDR KBG value makes this move obsolete midstock. This increase in damage and knockback will give Banjo another much-needed KO option (even if it's pretty late into the stock), while still retaining some of its combo potential
```diff
Specs:
+ Damage: 1.6/4.8->2.6/5.0
+ KBG: 100->136
```

## Damage Buffs
These moves lacked damage, to help make his throws slightly more versatile and to help Dtilt deal some more damage, these changes were made

### DTilt
```diff
Specs:
+ Damage: 6/5/4->8/6/6
```
### Fthrow
```diff
Specs:
+ Damage: 3/5->4/6
```
### Dthrow
```diff
Specs:
+ Damage: 6->8
```


## Fair Nerf
HDR Fair is good...too good. This tones down the KO power of the move as the rest of Banjos kit can now make up for it
```diff
Specs:
- KBG: 104 -> 94
```

CSharpM7 committed:
Cleaner implementation of Breegull Blaster

-Uses custom state to blend animation slightly better
-Uses a flag instead of int state machine
-Eggs shot are now preserved after bayonet state

CSharpM7 committed:
Merge branch 'HDR-Development:dev' into banjo-new-tricks

SuddyN commented on fighters/buddy/src/acmd/aerials.rs:
```diff
@@ -1,46 +1,114 @@
+        WorkModule::off_flag(fighter.module_accessor, *FIGHTER_STATUS_ATTACK_AIR_FLAG_ENABLE_LANDING);
+    }
+}
+#[acmd_script( agent = "buddy", script = "effect_attackairn", category = ACMD_EFFECT )]
+unsafe fn buddy_attack_air_n_effect(fighter: &mut L2CAgentBase) {
+    let lua_state = fighter.lua_state_agent;
+    frame(lua_state, 5.0);
+    if is_excute(fighter) {
+        LANDING_EFFECT_FLIP(fighter, Hash40::new("sys_spin_wind"), Hash40::new("sys_spin_wind"), Hash40::new("shoulderr"), 0, -2, 0, 0, 0, 0, 2.5, 0, 0, 0, 0, 0, 0, false, *EF_FLIP_NONE);
+    }
+    frame(lua_state, 17.0);
+    if is_excute(fighter) {
+        LANDING_EFFECT_FLIP(fighter, Hash40::new("sys_spin_wind"), Hash40::new("sys_spin_wind"), Hash40::new("shoulderr"), 0, -2, 0, 0, 0, 0, 2.5, 0, 0, 0, 0, 0, 0, false, *EF_FLIP_NONE);
+    }
+}
```
```suggestion
#[acmd_script( agent = "buddy", script = "effect_attackairn", category = ACMD_EFFECT )]
unsafe fn buddy_attack_air_n_effect(fighter: &mut L2CAgentBase) {
    let lua_state = fighter.lua_state_agent;
    frame(lua_state, 9.0);
    if is_excute(fighter) {
        EFFECT_FOLLOW_FLIP(fighter, Hash40::new("sys_spin_wind"), Hash40::new("sys_spin_wind"), Hash40::new("rot"), 0, 2.5, 0, 0, 180, 0, 1.5, true, *EF_FLIP_YZ);
        LAST_EFFECT_SET_RATE(fighter, 0.6);
    }
}
```
The current effect for NAir feels awkward. It leaves behind two wind trails which don't follow the actual trajectory of the backpack. My suggestion uses the same effect, but with `EFFECT_FOLLOW_FLIP` and a rate modifier. [Here is the result](https://user-images.githubusercontent.com/33226440/204070251-2fd17c9f-fd6c-4f8a-8019-2c2bfbcd5f10.mp4)

CSharpM7 replied:
Ya this does feel much more tight with the animation

SuddyN commented:
I tested these changes and they look good. Other than my comment above, everything looks great, and I'm excited to see this :)

CSharpM7 committed:
Update fighters/buddy/src/acmd/aerials.rs

Co-authored-by: Nathan Sudduth <33226440+SuddyN@users.noreply.github.com>

CSharpM7 committed:
Various Fixes

-Jab 1&2 now use 361 angle
-Vertical momentum ceased, which prevents air to ground transition during jump-squat
-Wall effect shrunken to .625 (or .5 for hitting a shield)
-kazooie uptilt lasts one frame longer, effect sped up
-Without golden feathers, Banjo no longer enters fail state during beakbomb unintentionally
-Red Feathers reset on idle during Training
-Beak Bomb slightly slower (x: 4.6->4.55)
-Beam Bomb exit speed multiplier removed

SkewedAskew committed:
Update opff.rs

SkewedAskew commented:
Notes:
- Up tilt has some notable z-axis issues as seen here:
![image](https://user-images.githubusercontent.com/55216313/209867481-cdee11fc-cc34-4448-bfe1-09ba60d27d87.png)
Additionally, Kazooie's effect trail was accidentally slowed down instead of sped up

- Beak Bomb's shield bounce seems to have gained some unintended momentum:
https://user-images.githubusercontent.com/55216313/209868440-74911d93-f932-4e51-aad7-f9acef73a92c.mp4

Otherwise looks and feels great implementation-wise

CSharpM7 committed:
Uptilt, Beakbomb Bonk, and Dash Attack Changes

-Kazooie Uptilt nolonger tracks a Kazooie bone, removing ZAxis shenanigans
-Effects speedup
-Dash attack rolls off stages, can be jump-canceled from frame 26 onward (per Wish's suggestion)
-Shieldbonking no longer changes kinetic type

CSharpM7 committed:
Merge branch 'banjo-new-tricks' of https://github.com/CSharpM7/HewDraw-Remix into banjo-new-tricks

SkewedAskew committed:
Update opff.rs

SkewedAskew reviewed (dismissed):
Lgtm, great work

CSharpM7 committed:
Prevents edge slip-off during WonderWing, causing invincible BeakBomb

Might find a cleaner work-around for this later

hdr-balance-backroom commented:
APPROVED

MadVillainLordQ reviewed (approved):
LGTM
